File: bridge/typescript/src/content.ts
// File content reading for MCP resource responses
// Handles text vs binary dispatch and path traversal guard.

import { readFileSync } from "node:fs";
import { resolve, join } from "node:path";
import type { KnowledgeUnit } from "./model.js";
import { isBinaryMime, resolveMimeType } from "./mapper.js";

export interface TextContent {
  type: "text";
  uri: string;
  mimeType: string;
  text: string;
}

export interface BlobContent {
  type: "blob";
  uri: string;
  mimeType: string;
  blob: string; // base64-encoded
}

export type ResourceContent = TextContent | BlobContent;

/**
 * Read the content of a KCP knowledge unit from disk.
 * Returns TextContent for text MIME types, BlobContent (base64) for binary.
 *
 * Throws if the path escapes the manifest root or the file does not exist.
 */
export function readUnitContent(
  manifestDir: string,
  unit: KnowledgeUnit,
  resourceUri: string
): ResourceContent {
  const manifestRoot = resolve(manifestDir);
  const resolved = resolve(join(manifestRoot, unit.path));

  // Security: reject path traversal
  if (!resolved.startsWith(manifestRoot + "/") && resolved !== manifestRoot) {
    throw new Error(`Path traversal rejected for unit '${unit.id}'`);
  }

  const mime = resolveMimeType(unit);

  if (isBinaryMime(mime)) {
    const data = readFileSync(resolved);
    return {
      type: "blob",
      uri: resourceUri,
      mimeType: mime,
      blob: data.toString("base64"),
    };
  } else {
    const text = readFileSync(resolved, "utf-8");
    return {
      type: "text",
      uri: resourceUri,
      mimeType: mime,
      text,
    };
  }
}

File: bridge/python/kcp_mcp/content.py
"""
File content reading for MCP resource serving.
Handles text/binary dispatch and path traversal guard.
"""
import base64
from pathlib import Path

from .mapper import is_binary_mime


class ResourceNotFoundError(Exception):
    pass


class PathTraversalError(Exception):
    pass


def read_resource_content(
    manifest_dir: Path,
    unit_path: str,
    mime: str,
) -> tuple[str, bool]:
    """
    Read a unit's file content from disk.

    Returns (content, is_binary) where:
    - is_binary=False: content is UTF-8 text
    - is_binary=True:  content is base64-encoded bytes

    Raises PathTraversalError if the resolved path escapes manifest_dir.
    Raises ResourceNotFoundError if the file does not exist.
    """
    resolved = (manifest_dir / unit_path).resolve()
    root = manifest_dir.resolve()

    try:
        resolved.relative_to(root)
    except ValueError:
        raise PathTraversalError(f"Path escapes manifest root: {unit_path!r}")

    if not resolved.exists():
        raise ResourceNotFoundError(f"Resource not found: {unit_path}")

    if is_binary_mime(mime):
        data = resolved.read_bytes()
        return base64.b64encode(data).decode("ascii"), True
    else:
        return resolved.read_text(encoding="utf-8"), False
